### ISSUE TITLE
First xdg-shell configure event is invalid

### ISSUE THREAD

**Issue by emersion** [swaywm/sway#2176]
When a client creates a new xdg-surface and commits a NULL buffer, we're supposed to send a configure event with the size of the window. We currently send it with a 0x0 size (which lets the client decides). We only send the first "real" configure event when the client maps (ie. after it has drawn its first frame).

We should send the correct configure event from the beginning.

**Comment by RyanDwyer** [comment]
My understanding is that the view can create itself but choose not to map immediately. During this delay the user might focus another container, in which case the size it needs could be different. So if we did try to preempt the size when it's created then we'd have to arrange it again when it sends the map request. The positive would be that when it maps, the surface will probably already be the correct size so the transaction would have one less view to wait on. This isn't a huge positive considering it still has to wait for the siblings to resize.

It would be fairly complicated to determine what size the view should be without actually arranging its siblings. You'd need a function similar to `arrange_children_of`. And to further complicate it, you don't know what size it should be until you know whether it'll have borders, will be placed on a different workspace due to criteria, or will be floating.

In my opinion the negatives greatly outweigh the positives. It's a lot of complicated code for something that the user is never going to see.

It sounds to me like the recommendation is intended for floating window managers.

**Comment by emersion** [comment]
>My understanding is that the view can create itself but choose not to map immediately.

I don't think so.

**Comment by RyanDwyer** [comment]
It must. After we send the configure, the client takes time to create the buffer, then commits or requests to be mapped when it's done. This could take 0ms, or it could take 10 seconds. It would be stupid for the compositor to block the event queue while waiting for the client to map, which means the user might change focus between these operations.

**Comment by emersion** [comment]
>It must. After we send the configure, the client takes time to create the buffer, then commits or requests to be mapped when it's done.

It's no different than a normal resize operation. If we block when resizing, we can block when mapping.

**Comment by SirCmpwn** [comment]
We can't let clients slowloris the compositor into a standstill.

**Comment by SirCmpwn** [comment]
Also, letting the client choose its initial size itself lets us stow that away and restore it when we float the client later.

**Comment by emersion** [comment]
>We can't let clients slowloris the compositor into a standstill.

I don't understand.

>Also, letting the client choose its initial size itself lets us stow that away and restore it when we float the client later.

The official way to do this is to send a 0x0 size.

**Comment by emersion** [comment]
To fix this, we should create a transaction when the client sends the initial NULL commit. We'd send a configure event with the right size even if the toplevel is currently unmapped. The transaction completes when the client sends an ack_configure and commits a buffer.

**Comment by anarcat** [comment]
Just wanting to add that gstreamer has a good reproducer for this:

```
gst-launch-1.0 videotestsrc ! waylandsink  fullscreen=TRUE
```

That crashes completely here, which is probably suboptimal from gstreamer, but that's another story:

```
ERROR: from element /GstPipeline:pipeline0/GstWaylandSink:waylandsink0: Window has no size set
```

Here's the full log:

```
anarcat@angela:~$ WAYLAND_DEBUG=1 gst-launch-1.0 -v videotestsrc ! waylandsink  fullscreen=TRUE

(gst-launch-1.0:1786375): GStreamer-Wayland-CRITICAL **: 15:57:26.538: gst_wl_window_ensure_fullscreen: assertion 'self' failed
Setting pipeline to PAUSED ...
[4283479.379]  -> wl_display@1.get_registry(new id wl_registry@2)
[4283479.396]  -> wl_display@1.sync(new id wl_callback@3)
[4283479.609] wl_display@1.delete_id(3)
[4283479.620] wl_registry@2.global(1, "wl_shm", 1)
[4283479.631]  -> wl_registry@2.bind(1, "wl_shm", 1, new id [unknown]@4)
[4283479.640] wl_registry@2.global(2, "wl_drm", 2)
[4283479.648] wl_registry@2.global(3, "zwp_linux_dmabuf_v1", 4)
[4283479.655]  -> wl_registry@2.bind(3, "zwp_linux_dmabuf_v1", 1, new id [unknown]@5)
[4283479.663] wl_registry@2.global(4, "wl_compositor", 5)
[4283479.669]  -> wl_registry@2.bind(4, "wl_compositor", 4, new id [unknown]@6)
[4283479.676] wl_registry@2.global(5, "wl_subcompositor", 1)
[4283479.682]  -> wl_registry@2.bind(5, "wl_subcompositor", 1, new id [unknown]@7)
[4283479.688] wl_registry@2.global(6, "wl_data_device_manager", 3)
[4283479.693] wl_registry@2.global(7, "zwlr_gamma_control_manager_v1", 1)
[4283479.704] wl_registry@2.global(8, "zxdg_output_manager_v1", 3)
[4283479.713] wl_registry@2.global(9, "org_kde_kwin_idle", 1)
[4283479.718] wl_registry@2.global(10, "ext_idle_notifier_v1", 1)
[4283479.727] wl_registry@2.global(11, "zwp_idle_inhibit_manager_v1", 1)
[4283479.736] wl_registry@2.global(12, "zwlr_layer_shell_v1", 4)
[4283479.745] wl_registry@2.global(13, "xdg_wm_base", 2)
[4283479.754]  -> wl_registry@2.bind(13, "xdg_wm_base", 1, new id [unknown]@8)
[4283479.764] wl_registry@2.global(14, "zwp_tablet_manager_v2", 1)
[4283479.771] wl_registry@2.global(15, "org_kde_kwin_server_decoration_manager", 1)
[4283479.780] wl_registry@2.global(16, "zxdg_decoration_manager_v1", 1)
[4283479.789] wl_registry@2.global(17, "zwp_relative_pointer_manager_v1", 1)
[4283479.796] wl_registry@2.global(18, "zwp_pointer_constraints_v1", 1)
[4283479.802] wl_registry@2.global(19, "wp_presentation", 1)
[4283479.810] wl_registry@2.global(20, "zwlr_output_manager_v1", 4)
[4283479.818] wl_registry@2.global(21, "zwlr_output_power_manager_v1", 1)
[4283479.827] wl_registry@2.global(22, "zwp_input_method_manager_v2", 1)
[4283479.836] wl_registry@2.global(23, "zwp_text_input_manager_v3", 1)
[4283479.844] wl_registry@2.global(24, "zwlr_foreign_toplevel_manager_v1", 3)
[4283479.853] wl_registry@2.global(25, "ext_session_lock_manager_v1", 1)
[4283479.860] wl_registry@2.global(26, "wp_drm_lease_device_v1", 1)
[4283479.868] wl_registry@2.global(27, "zwlr_export_dmabuf_manager_v1", 1)
[4283479.878] wl_registry@2.global(28, "zwlr_screencopy_manager_v1", 3)
[4283479.887] wl_registry@2.global(29, "zwlr_data_control_manager_v1", 2)
[4283479.893] wl_registry@2.global(30, "zwp_primary_selection_device_manager_v1", 1)
[4283479.902] wl_registry@2.global(31, "wp_viewporter", 1)
[4283479.910]  -> wl_registry@2.bind(31, "wp_viewporter", 1, new id [unknown]@9)
[4283479.923] wl_registry@2.global(32, "wp_single_pixel_buffer_manager_v1", 1)
[4283479.934] wl_registry@2.global(33, "zxdg_exporter_v1", 1)
[4283479.943] wl_registry@2.global(34, "zxdg_importer_v1", 1)
[4283479.949] wl_registry@2.global(35, "zxdg_exporter_v2", 1)
[4283479.958] wl_registry@2.global(36, "zxdg_importer_v2", 1)
[4283479.966] wl_registry@2.global(37, "xdg_activation_v1", 1)
[4283479.974] wl_registry@2.global(38, "zwp_virtual_keyboard_manager_v1", 1)
[4283479.983] wl_registry@2.global(39, "zwlr_virtual_pointer_manager_v1", 2)
[4283479.992] wl_registry@2.global(40, "zwlr_input_inhibit_manager_v1", 1)
[4283479.998] wl_registry@2.global(41, "zwp_keyboard_shortcuts_inhibit_manager_v1", 1)
[4283480.007] wl_registry@2.global(42, "wl_seat", 8)
[4283480.018] wl_registry@2.global(43, "zwp_pointer_gestures_v1", 3)
[4283480.027] wl_registry@2.global(44, "wl_output", 4)
[4283480.036] wl_callback@3.done(86791)
[4283480.046]  -> wl_display@1.sync(new id wl_callback@3)
[4283480.301] wl_display@1.delete_id(3)
[4283480.311] wl_shm@4.format(0)
[4283480.318] wl_shm@4.format(1)
[4283480.326] wl_shm@4.format(875709016)
[4283480.333] wl_shm@4.format(875708993)
[4283480.341] wl_shm@4.format(875710274)
[4283480.349] wl_shm@4.format(842094674)
[4283480.355] wl_shm@4.format(842088786)
[4283480.362] wl_shm@4.format(892426322)
[4283480.370] wl_shm@4.format(892420434)
[4283480.377] wl_shm@4.format(909199186)
[4283480.384] wl_shm@4.format(808665688)
[4283480.391] wl_shm@4.format(808665665)
[4283480.398] wl_shm@4.format(1211384408)
[4283480.405] wl_shm@4.format(1211384385)
[4283480.412] wl_shm@4.format(942948952)
[4283480.419] wl_shm@4.format(942948929)
[4283480.426] zwp_linux_dmabuf_v1@5.format(1211384385)
[4283480.434] zwp_linux_dmabuf_v1@5.format(1211384408)
[4283480.449] zwp_linux_dmabuf_v1@5.format(942948929)
[4283480.455] zwp_linux_dmabuf_v1@5.format(942948952)
[4283480.462] zwp_linux_dmabuf_v1@5.format(808669761)
[4283480.469] zwp_linux_dmabuf_v1@5.format(808669784)
[4283480.476] zwp_linux_dmabuf_v1@5.format(808665665)
[4283480.483] zwp_linux_dmabuf_v1@5.format(875713089)
[4283480.491] zwp_linux_dmabuf_v1@5.format(875708993)
[4283480.499] zwp_linux_dmabuf_v1@5.format(875713112)
[4283480.506] zwp_linux_dmabuf_v1@5.format(875709016)
[4283480.514] zwp_linux_dmabuf_v1@5.format(892424769)
[4283480.521] zwp_linux_dmabuf_v1@5.format(909199186)
[4283480.528] zwp_linux_dmabuf_v1@5.format(538982482)
[4283480.535] zwp_linux_dmabuf_v1@5.format(540422482)
[4283480.542] zwp_linux_dmabuf_v1@5.format(943215175)
[4283480.550] zwp_linux_dmabuf_v1@5.format(842224199)
[4283480.558] zwp_linux_dmabuf_v1@5.format(961959257)
[4283480.565] zwp_linux_dmabuf_v1@5.format(825316697)
[4283480.580] zwp_linux_dmabuf_v1@5.format(842093913)
[4283480.588] zwp_linux_dmabuf_v1@5.format(909202777)
[4283480.596] zwp_linux_dmabuf_v1@5.format(875713881)
[4283480.603] zwp_linux_dmabuf_v1@5.format(961893977)
[4283480.614] zwp_linux_dmabuf_v1@5.format(825316953)
[4283480.622] zwp_linux_dmabuf_v1@5.format(842094169)
[4283480.629] zwp_linux_dmabuf_v1@5.format(909203033)
[4283480.636] zwp_linux_dmabuf_v1@5.format(875714137)
[4283480.643] zwp_linux_dmabuf_v1@5.format(842094158)
[4283480.651] zwp_linux_dmabuf_v1@5.format(808530000)
[4283480.658] zwp_linux_dmabuf_v1@5.format(842084432)
[4283480.665] zwp_linux_dmabuf_v1@5.format(909193296)
[4283480.674] zwp_linux_dmabuf_v1@5.format(909203022)
[4283480.681] zwp_linux_dmabuf_v1@5.format(1448433985)
[4283480.688] zwp_linux_dmabuf_v1@5.format(1448434008)
[4283480.696] zwp_linux_dmabuf_v1@5.format(808531033)
[4283480.702] zwp_linux_dmabuf_v1@5.format(842085465)
[4283480.710] zwp_linux_dmabuf_v1@5.format(909194329)
[4283480.717] zwp_linux_dmabuf_v1@5.format(1448695129)
[4283480.725] zwp_linux_dmabuf_v1@5.format(1498831189)
[4283480.735] zwp_linux_dmabuf_v1@5.format(808530521)
[4283480.743] zwp_linux_dmabuf_v1@5.format(842084953)
[4283480.751] zwp_linux_dmabuf_v1@5.format(909193817)
[4283480.759] wl_callback@3.done(86791)
Pipeline is PREROLLING ...
/GstPipeline:pipeline0/GstVideoTestSrc:videotestsrc0.GstPad:src: caps = video/x-raw, format=(string)BGRA, width=(int)320, height=(int)240, framerate=(fraction)30/1, multiview-mode=(string)mono, pixel-aspect-ratio=(fraction)1/1, interlace-mode=(string)progressive
/GstPipeline:pipeline0/GstWaylandSink:waylandsink0.GstPad:sink: caps = video/x-raw, format=(string)BGRA, width=(int)320, height=(int)240, framerate=(fraction)30/1, multiview-mode=(string)mono, pixel-aspect-ratio=(fraction)1/1, interlace-mode=(string)progressive
[4283484.795]  -> wl_compositor@6.create_surface(new id wl_surface@3)
[4283484.826]  -> wl_compositor@6.create_surface(new id wl_surface@10)
[4283484.832]  -> wl_subcompositor@7.get_subsurface(new id wl_subsurface@11, wl_surface@10, wl_surface@3)
[4283484.838]  -> wl_subsurface@11.set_desync()
[4283484.843]  -> wp_viewporter@9.get_viewport(new id wp_viewport@12, wl_surface@3)
[4283484.849]  -> wp_viewporter@9.get_viewport(new id wp_viewport@13, wl_surface@10)
[4283484.854]  -> wl_compositor@6.create_region(new id wl_region@14)
[4283484.858]  -> wl_surface@10.set_input_region(wl_region@14)
[4283484.863]  -> wl_region@14.destroy()
[4283484.868]  -> xdg_wm_base@8.get_xdg_surface(new id xdg_surface@15, wl_surface@3)
[4283484.874]  -> xdg_surface@15.get_toplevel(new id xdg_toplevel@16)
[4283484.882]  -> xdg_toplevel@16.set_fullscreen(nil)
[4283484.887]  -> wl_surface@3.commit()
[4283491.140] wl_display@1.delete_id(14)
[4283491.186] xdg_wm_base@8.ping(86793)
[4283491.193]  -> xdg_wm_base@8.pong(86793)
[4283491.204] xdg_toplevel@16.configure(0, 0, array[0])
[4283491.212] xdg_surface@15.configure(86792)
[4283491.217]  -> xdg_surface@15.ack_configure(86792)
[4283491.382]  -> wl_surface@3.commit()
ERROR: from element /GstPipeline:pipeline0/GstWaylandSink:waylandsink0: Window has no size set
Additional debug info:
../ext/wayland/gstwaylandsink.c(926): gst_wayland_sink_show_frame (): /GstPipeline:pipeline0/GstWaylandSink:waylandsink0:
Make sure you set the size after calling set_window_handle
ERROR: pipeline doesn't want to preroll.
ERROR: from element /GstPipeline:pipeline0/GstVideoTestSrc:videotestsrc0: Internal data stream error.
Additional debug info:
../libs/gst/base/gstbasesrc.c(3132): gst_base_src_loop (): /GstPipeline:pipeline0/GstVideoTestSrc:videotestsrc0:
streaming stopped, reason error (-5)
ERROR: pipeline doesn't want to preroll.
Setting pipeline to NULL ...
[4283491.926]  -> xdg_toplevel@16.destroy()
[4283491.950]  -> xdg_surface@15.destroy()
[4283491.956]  -> wp_viewport@13.destroy()
[4283491.959]  -> wl_subsurface@11.destroy()
[4283491.963]  -> wl_surface@10.destroy()
[4283491.967]  -> wp_viewport@12.destroy()
[4283491.971]  -> wl_surface@3.destroy()
[4283492.196]  -> wp_viewporter@9.destroy()
[4283492.218]  -> zwp_linux_dmabuf_v1@5.destroy()
[4283492.223]  -> xdg_wm_base@8.destroy()
[4283492.228]  -> wl_subcompositor@7.destroy()
Freeing pipeline ...
```
